# Clase 14/02/2023 #

## Resumen ##

### Manejo de Excepciones ##

El dia de hoy empezamos viendo el manejo de excepciones primero hicimos un pequeño programa que divide dos numeros al hacer la division entre 0 nos arroja un error.

![image](https://user-images.githubusercontent.com/123017277/218897478-25287cfa-7e72-419e-9990-d887d0091957.png)

El flujo de la division por 0 es un flujo no comun hay algunas cosas que podemos hacer para que deje de arrojar el error:

lo primero que podemos hacer es agregar un if el cual si ingresamos un 0 como divisor va a regresar un mensaje mucho mejor y mas entendible para el usuario.

![image](https://user-images.githubusercontent.com/123017277/218898608-9fb9f1c3-2f00-4f13-8e02-626d4ea0e838.png)

con esto se mejora un poco pero sigue siendo incomodo ya que la division no es 0 por lo que necesitamos una mejor solucion por lo que vamos a usar el metodo try catch.

![image](https://user-images.githubusercontent.com/123017277/218899286-cb6400af-2417-4c43-9ef6-1914989a8ea9.png)

con esto obtenemos un manejador de errores que se parece mucho al if de la imagen anterior pero con una estructura diferente llamada try catch, try quiere decir intenta lo que traes dentro  si no es posible pasa a la seccion de catch.

![image](https://user-images.githubusercontent.com/123017277/218899746-67d32cdf-191e-4a8a-9c90-4e19c3db502a.png)

En e quedan guardados los errores que java nos presentaria si no hacemos algo, esto nos da la opcion de pintarlos si queremos o la opcion de preguntar que tiene dentro e y hacer algo al respecto o si ya tenemos claro que es lo que hay guardado en e lo podemos ignorar y hacer lo que tenemos que hacer.

La diferencia entre trycatch e if else es que try catch ademas de que te ayuda a manejar el error te estructura lo suficientemente bien como para poder tener acceso al error y si ueremos podemos hacer preguntas.

ya que en la parte de try podemos poner todas las condiciones que queramos y si alguna de ellas genera un error va a saltarse a la parte de catches algo que no teniamos con el if ya que if solo evalua una condicion.

![image](https://user-images.githubusercontent.com/123017277/218905516-2ad3a707-b5ad-4487-92e9-18be82d26b7a.png)

para demostrar lo anterior corremos el programa en la forma de debug 

![exception](https://user-images.githubusercontent.com/123017277/218906178-eb3e1e03-26e0-4b80-bbd1-e44ca21e68ae.gif)

en este gif podemos ver que el programa corre perfectamente pero si ingresamos un cero en la variable i al llegar al allinea 19 se genera un error y se brinca de la linea 19 a la 22 al metodo catch.

![exception2](https://user-images.githubusercontent.com/123017277/218908093-f70bcd0e-4362-4145-b588-640f7a91c637.gif)

nos imprime

![image](https://user-images.githubusercontent.com/123017277/218908272-67b5ffdc-d056-47b3-a310-a18325dd8a8c.png)

un ejemplo con otro tipo de error ingresamos en la linea 18 una excepcion creada por nosotros 

![image](https://user-images.githubusercontent.com/123017277/218909013-41789f1c-16b7-4400-8528-194b822afcc7.png)

al correr el programa en forma debug cuandoo corre la linea18 se genera la excepcion y salta a la linea 23

![image](https://user-images.githubusercontent.com/123017277/218909640-76c297ce-0b24-4ea7-bc64-86c6ee21f71d.png)

![image](https://user-images.githubusercontent.com/123017277/218909753-a4e8eec7-d42c-4aac-8bff-7706a43c3256.png)

Ahora intentamos correr con otro error le pedimos que lea un archivo que no existe en el equipo, cuando corremos el programa llega a la linea 23 y de ahi se salta a la linea 30 donde podemos ver el error que se genero.

![1](https://user-images.githubusercontent.com/123017277/218911700-926b2c15-4f9d-4027-a18a-f17241ada183.gif)

![image](https://user-images.githubusercontent.com/123017277/218916315-21ed924a-8e3b-4e40-9973-b97fbed82901.png)

Esta excepcion tiene una propiedad llamada file con esta propiedad podriamos decirle get message para mejorar nuestro programa

![image](https://user-images.githubusercontent.com/123017277/218916501-bda0f841-972c-4091-81ff-a818ea93ed31.png)

![get](https://user-images.githubusercontent.com/123017277/218917161-3fa0fe43-a5a7-4fd3-96d5-0f74bdb09a84.gif)

De esta manera podemos manipular el mensaje de salida indicando incluso cosas muy particulares del error.

Tambien podemos generar una excepcion cuando no queremos que pasen algunas cosas por ejemplo si no queremos que se realicen operaciones con numeros negativos.

![image](https://user-images.githubusercontent.com/123017277/218917908-419351a4-35af-455e-bd38-ff7090e60069.png)

Las excepciones son un medio que nos ofrece java para poder gestionar aquellos eventos que hacen que mi flujo no sea el esperado.

Creamos un clase llamada NoNegativenumbers exception en la cual creamos un construtor el cual tiene como caracteristica que tienen el mimsmo nombre de la clase y no retorna nada a demas agregamos un super para llamar al mensaje "No se alen numeros negativos".

![image](https://user-images.githubusercontent.com/123017277/218920414-dd6de399-8389-4742-87e9-b540b66f4505.png)

Entonces podemos mandar a llamar nuestra excepcion al correr el programa en la linea 23 nos arroja como resultado nuestra excepcion

![image](https://user-images.githubusercontent.com/123017277/218921460-f660204f-84ab-4085-bd8e-f848fb543e16.png)

Catch puede aceptar un objeto tipo excepcion o un desendiente de ella, de est aforma podemos crear nuestras propias excepciones para los flujos alternativos de nuestro programa.

Las distintas excepciones que dispara nuestro programa las podemos meter en diferentes bloques de catch o bien podemos tener un solo catch con una clase que sea el padre de todos y se manda a llamar un mensaje que pertenezca a todos ellos no es eficiente ya que no se puede ser muy claro hacia el usuario.

![image](https://user-images.githubusercontent.com/123017277/218928130-d74fb475-e9f4-4eda-af69-669e5f4f42bb.png)
![image](https://user-images.githubusercontent.com/123017277/218928456-e0402a8d-1339-4759-a0f2-e2b62237b9ec.png)

Si quitamos el try catch del programa podemos cambiarlo por throws

![image](https://user-images.githubusercontent.com/123017277/218929122-23de2bba-e1bd-4227-92d2-3dbdcd9d09a2.png)

y poner un solo try catch a divide

![image](https://user-images.githubusercontent.com/123017277/218929348-92bb9b1f-5563-4a91-ac10-e50c2c979ce8.png)

La moraleja es podemos controlar los flujos alternativos del programa de manera efciente y elegante, la forma es poner el flujo normal en el try y tener los catch que sean para atender los excepciones que se van disparando a lo largo de nuestro programa.

![image](https://user-images.githubusercontent.com/123017277/218932219-ff86334c-f2b7-45a8-804e-51fc56cd1efa.png)

Como en este programa tuvimos 3 excepciones y se agregaron 3 catch para atenderlas.

Tambien podemos diferir algunos catch alos metodos que los mandan llamar podemos crear toda un serie de suceciones de llamadas try catch ara que se vayan resolviendo poco a poco

![image](https://user-images.githubusercontent.com/123017277/218934759-6b807262-e12c-401e-8057-0fb985ade854.png)

Un pequeño diagrama en excel con las altertivas para el tratamiento de las excepciones

![image](https://user-images.githubusercontent.com/123017277/218935228-b5fb81de-d4e1-4463-b717-3fe833ad3b04.png)

### Colecciones ###

empezamos desde cero creando na nueva clase lammada prueba 5 

![image](https://user-images.githubusercontent.com/123017277/218946939-9972c8a6-7ae2-4a3b-813c-1f4626d5ddd0.png)

creamos un array de tipo Rana lo cual no es puy viable ya que se gasta memoria y cuabdo llegue al limite nos arrojaria un error

![image](https://user-images.githubusercontent.com/123017277/218951579-ce543641-6b3e-4978-8820-650b6a3c8eb4.png)

Java nos va a permitir seguir metiendo objetos ya que no es un array si no un objeto de tipo list

![image](https://user-images.githubusercontent.com/123017277/218952396-8ac80870-846e-4f65-9998-2080cbd10b35.png)

se llaman colecciones ya que hay muchos objetos como el list, Set, Hash, etc. Que nos permiten guardar muchas cosas dentro sin restricion

list es una interface.

![image](https://user-images.githubusercontent.com/123017277/218953115-92c242f4-379b-430f-b044-32631b85858f.png)

¿Porque hay tres intefaces seguidas si con una sola se podria tener todo?

![image](https://user-images.githubusercontent.com/123017277/218953548-ac7588dd-00fe-458a-8d9b-0d9f2c88e174.png)

Debido a que hay arboles de herencias es necesario ir agregando funcionalidades a las interfaces que nos interesan.

a Array list puede guardar de todo tipo de Objetos lo cual es muy lamentable ya que se pierde el orden para solucionar esto se creo una forma en la cual podemos decirle a la bolsa que tipo de objetos va a guardar y solo va a aceptar ese tipo de objetos.

![image](https://user-images.githubusercontent.com/123017277/218955365-840d572d-e0ae-4a37-a65a-e833f8374d47.png)
![image](https://user-images.githubusercontent.com/123017277/218955460-6ee5eb84-592c-4c58-b5df-d8af54a7fc9c.png)

solo se pueden ingresar el objeto indicado o descendientes de este.
